### ISSUE TITLE
Enable and disable stateful precompiles using upgradeBytes

### ISSUE THREAD

**Issue by darioush** [ava-labs/subnet-evm#154]
## Context
SubnetEVM needs to handle upgradeBytes in order to allow subnetEVM users to easily upgrade an instance of the vanilla subnetEVM without creating their own fork.

## Upgrade Bytes Serialization
Stateful precompiles are enabled in subnetEVM by placing a config in the ChainConfig. However, this locks blockchains in to that config for their entire lifetime and we would like to support toggling stateful precompiles as a sometimes necessary (although still discouraged and to be used as infrequently as possible) capability.

upgradeBytes will be specified as a json object that may contain the following keys:
```
{
  “precompileUpgrades”: [...]
  “networkUpgrades”: {
     “SubnetEVMTimestamp”: 0
   }
}
```

Suggested approach: If `networkUpgrades` is specified, it is considered a full replacement for the timestamp (or block #) that activates forks specified in the chain config. Therefore any activated forks must be specified or `upgradeBytes` will be rejected as incompatible
_Alternative 1: only apply upgrades specified in `upgradeBytes`. This provides an easier UX, where specifying all upgrades is more explicit._
_Alternative 2: make `networkUpgrades` mandatory._

Suggested approach: If `precompileUpgrades` is specified, it contains an array of possible precompile configs. Each precompile config must include a `blockTimestamp`, and `blockTimestamp`s specified in the array must be sorted in increasing order (duplicate timestamps allowed). If a precompile upgrade specifies `"disable": true`, it will deactivate the precompile (that may be later re-activated). eg,

```
"precompileUpgrades": [
        {
            "contractNativeMinterConfig": {
                "blockTimestamp": 1657892836,
                "disable": true
            }
        },
        {
            "contractNativeMinterConfig": {
                "blockTimestamp": 1657893137,
                "adminAddresses": [
                  "0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC"
                ]
              }
        }
      ]
```
The benefit of this approach is reusing the same configuration types in precompileUpgrades as in genesis.

Alternative:
Use a map like `“timestamp”: [ {change} ]` then overriding `blockTimestamp` in each config. Downside will be special handling of `omitempty` for blockTimestamp in genesis vs. upgradeBytes.

Alternative 2:
Special parsing for Unmarshalling JSON into an array of configs with special case handling when there is only one present vs. an actual array for backwards compatibility. This is a little gross, so very open to alternatives here, but not sure what the best path forward is.

## DB Spec and Startup
On startup, the VM is passed in `genesisBytes` and `upgradeBytes`. The `genesisBytes` passed into the VM will always be the same throughout the lifetime of the chain.

`upgradeBytes` should be used to configure the timing of network upgrades without changing the code.

When subnetEVM starts up, it should read in the `upgradeBytes` and verify that it is compatible with the genesis chain config and previously applied `upgradeBytes` that is currently on disk. If it’s not, this should return a fatal error.

If the config is compatible with what’s on disk, then it may contain a new upgrade that is not present in the genesis chain config and previously applied `upgradeBytes` stored on disk. In this case, the `upgradeBytes` should be validated with `CheckConfigForkOrder` and overwrite what is currently on disk.

Where to store upgradeBytes?
Suggestion: Directly into a key in `vm.db` (eg, `upgrade_bytes`). Seems simple.

There are a few cases that we will want to make sure to handle:
- Compatible Upgrade that Occurs in the Future: If upgradeBytes contains a new upgrade that occurs at a future timestamp relative to the last accepted block, then it should overwrite the chain config stored on disk and go into effect
After this, we should require that any upgradeBytes that were supplied once are always supplied in the same way
- Incompatible upgrade that occurs in the future
If upgradeBytes contains a new upgrade that occurs in the future but is not compatible with the current chain config, then it should return a fatal error
- Attempt to perform an upgrade prior to the last accepted block’s timestamp
Any new upgrade that occurs prior to the timestamp of the last accepted block should return a fatal error
- Delete upgrade that occurs in the future
If there is a future upgrade stored on disk that has not activated yet, it is considered abortable.
If upgradeBytes is set with an abortable upgrade removed from the parsed ChainConfig the chain will remove the abortable upgrade from the ChainConfig on startup
- Delete upgrade that has already occurred
Will be considered a fatal error.

## Enabling and disabling pre-compiles
When a new block is processed:
- Check if any upgrades are being enabled or disabled
- For all transitions that are occurring from the previous block to the current block:
  * If the upgrade enables a precompile:
     - Set the nonce for the precompile to 1
     - Set the code of the precompile to non-empty
     - Perform additional precompile specific state modifications
  * If the upgrade disables a precompile:
     - Suicide the address space for the precompile to prevent leaving certain allow-listed addresses in place to avoid any unintended consequences.

## Enabling and disabling other features?
Open question: If we want to support changing allowFeeRecepients in upgrade bytes do we…
Support this as a fork with a blockTimestamp (then it cannot be disabled)
Support this as a precompile configuration
What details will be needed for this?

**Comment by ceyonur** [comment]
> Suggested approach: If networkUpgrades is specified, it is considered a full replacement for the timestamp (or block #) that activates forks specified in the chain config. Therefore any activated forks must be specified or upgradeBytes will be rejected as incompatible
Alternative 1: only apply upgrades specified in upgradeBytes. This provides an easier UX, where specifying all upgrades is more explicit.
Alternative 2: make networkUpgrades mandatory.

I'd prefer Alternative 1 because if we add a new precompile/network upgrade the existing upgradeBytes should keep working. 

> Alternative 1:
Use a map like “timestamp”: [ {change} ] then overriding blockTimestamp in each config. Downside will be special handling of omitempty for blockTimestamp in genesis vs. upgradeBytes.
Alternative 2:
Special parsing for Unmarshalling JSON into an array of configs with special case handling when there is only one present vs. an actual array for backwards compatibility. This is a little gross, so very open to alternatives here, but not sure what the best path forward is.

I think we can ditch Alternative 1 here and keep the existing structure. I think we should not let multiple structs in chain config. We should keep chainConfig as it is (with single config struct) and use array of configs for the upgradeBytes. So Alternative 2 would not be necessary.

**Comment by aaronbuchwald** [comment]
- Let's store the `upgrade config` alongside the `chainConfig` to keep them in the same place and make it easier to check them alongside each other.

> upgradeBytes will be specified as a json object that may contain the following keys:
> 
> ```
> {
>   “precompileUpgrades”: [...]
>   “networkUpgrades”: {
>      “SubnetEVMTimestamp”: 0
>    }
> }
> ```
> 
> Suggested approach: If `networkUpgrades` is specified, it is considered a full replacement for the timestamp (or block #) that activates forks specified in the chain config. Therefore any activated forks must be specified or `upgradeBytes` will be rejected as incompatible _Alternative 1: only apply upgrades specified in `upgradeBytes`. This provides an easier UX, where specifying all upgrades is more explicit._ _Alternative 2: make `networkUpgrades` mandatory._

I'd prefer to make it mandatory that each upgrade in `networkUpgrades` specifies the full config. For example,

if the genesis contains:
```
{
    "subnetEVMTimestamp": 0
}
```

and we want to include the upgrade subnetEVM2Timestamp, the first and only item in `networkUpgrades` would then be:

```
{
    "networkUpgrades": [
        {
            "subnetEVMTimestamp": 0,
            "subnetEVM2Timestamp": 1658239000
        }
    ]
}
```

which we can verify is correct via `CheckCompatible`: https://github.com/ava-labs/subnet-evm/blob/master/params/config.go#L226.

For subsequent upgrades, we can verify each upgrade on top of the next. If we want to add `subnetEVM3Timestamp` at a later date, then it would look like this:

```
{
    "networkUpgrades": [
        {
            "subnetEVMTimestamp": 0,
            "subnetEVM2Timestamp": 1658239000
        },
        {
            "subnetEVMTimestamp": 0,
            "subnetEVM2Timestamp": 1658239000,
            "subnetEVM3Timestamp": 1660239000
        }
    ]
}
```

**Comment by dasconnor** [comment]
> Suggested approach: If precompileUpgrades is specified, it contains an array of possible precompile configs. Each precompile config must include a blockTimestamp, and blockTimestamps specified in the array must be sorted in increasing order (duplicate timestamps allowed). If a precompile upgrade specifies "disable": true, it will deactivate the precompile (that may be later re-activated).

Where will these need to be sorted? Does the user need to provide them in sorted order or can we parse user input in any order and sort it ourselves?

Also what smart contract accessible methods do you plan to expose?

**Comment by aaronbuchwald** [comment]
> > Suggested approach: If precompileUpgrades is specified, it contains an array of possible precompile configs. Each precompile config must include a blockTimestamp, and blockTimestamps specified in the array must be sorted in increasing order (duplicate timestamps allowed). If a precompile upgrade specifies "disable": true, it will deactivate the precompile (that may be later re-activated).
> 
> Where will these need to be sorted? Does the user need to provide them in sorted order or can we parse user input in any order and sort it ourselves?
> 

When the upgrades are passed in to the VM they will need to be sorted by timestamp. For the avalanche-cli, this means that it can sort them on behalf of the user.

> Also what smart contract accessible methods do you plan to expose?

The precompiles themselves will be accessible by smart contracts. The configuration of these will be up to the VM and will not be accessible by smart contracts in any way.

**Comment by dasconnor** [comment]
I guess I'm a little confused. If this precompile is activated, what is its interface? The user can't provide the upgrades he wants in the genesis (because he doesn't know them at that time), so how does he submit them? 

For example, I want to turn off tx whitelisting at some point in time, but I don't know when. When I do decide I'm ready, how do I activate that feature? I thought it would be via a smart contract call.

**Comment by darioush** [comment]
> For example, I want to turn off tx whitelisting at some point in time, but I don't know when. When I do decide I'm ready, how do I activate that feature? I thought it would be via a smart contract call.

The upgrade is performed when avalanchego passes "upgradeBytes" to the VM on initialization. Like other network upgrades, this requires coordinating nodes to add the relevant upgradeBytes to their node and restart it before the timestamp specified in the upgradeBytes.

The bytes are specified in `upgrade.json` see [here](https://github.com/ava-labs/avalanchego/blob/547bdd34f9d3fa438eeaaf44efa865a9e2806148/config/config.go#L984-L994)

**Comment by aaronbuchwald** [comment]
Resolved by #148
